src/v/kafka/requests/fetch_request.cc

```diff
@@ -428,92 +428,112 @@ handle_ntp_fetch(op_context& octx, model::ntp ntp, fetch_config config) {
           try {
               auto response = f.get0();
               response.id = p_id;
-              octx.add_partition_response(std::move(response));
+              octx.set_partition_response(std::move(response));
           } catch (...) {
               /*
                * TODO: this is where we will want to handle any storage
                * specific errors and translate them into kafka response
                * error codes.
                */
-              octx.response_error = true;
-              octx.add_partition_response(make_partition_response_error(
+              octx.set_partition_response(make_partition_response_error(
                 error_code::unknown_server_error));
           }
       });
 }
 
+static ss::future<> fetch_topic_partition(
+  op_context& octx, const fetch_request::const_iterator::value_type& p) {
+    /*
+     * the next topic-partition to fetch
+     */
+    auto& topic = *p.topic;
+    auto& part = *p.partition;
+    if (p.new_topic && octx.initial_fetch) {
+        octx.start_response_topic(topic);
+    }
+
+    if (
+      model::timeout_clock::now() > octx.deadline.value_or(model::no_timeout)) {
+        octx.set_partition_response(
+          make_partition_response_error(error_code::request_timed_out));
+        return ss::now();
+    }
+
+    // if over budget create placeholder response
+    if (octx.bytes_left <= 0) {
+        octx.set_partition_response(
+          make_partition_response_error(error_code::message_too_large));
+        return ss::now();
+    }
+    // if we already have data in response for this partition skip it
+    if (!octx.initial_fetch) {
+        auto& partition_response = octx.response_iterator->partition_response;
+        // do not try to fetch again if partition response is already there i.e.
+        // it has data or
+        if (
+          partition_response->error != error_code::none
+          || !partition_response->record_set->empty()) {
```

Comment:
what happens if a partition has some data (so we skip it here) but overall we don't meet the threshold for returning to the client before a timeout expires. might we skip re-fetching a partition that would let us meet the min bytes threshold?

Reply:
That indeed would be the problem i.e. when partition contain some data but the threshold isn't met we will return a timeout. I I will allow skipping re-fetching a partition only when the threshold is met (in general)